src/test/java/com/google/api/codegen/GapicTestBase.java

```diff
@@ -0,0 +1,197 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+/**
+ * Base class for code generator baseline tests.
+ */
+public abstract class GapicTestBase extends ConfigBaselineTestCase {
+
+  private static final Pattern BASELINE_PATTERN = Pattern.compile("((?:-|\\w)+)\\[((?:-|\\w)+)\\]");
+
+  // Wiring
+  // ======
+
+  private final String name;
+  private final String idForFactory;
+  private final String[] gapicConfigFileNames;
+  private final String snippetName;
+  protected ConfigProto config;
+
+  public GapicTestBase(
+      String name,
+      String idForFactory,
+      String[] gapicConfigFileNames,
+      String snippetName) {
+    this.name = name;
+    this.idForFactory = idForFactory;
+    this.gapicConfigFileNames = gapicConfigFileNames;
+    this.snippetName = snippetName;
+  }
+
+  @Override
+  protected Object run() {
+    List<GeneratedResult> results = generate();
+    Truth.assertThat(results).isNotNull();
+
+    ImmutableMap.Builder<String, Doc> builder = new ImmutableMap.Builder<String, Doc>();
+    for (GeneratedResult result : results) {
+      builder.put(result.getFilename(), result.getDoc());
+    }
+    return builder.build();
+  }
+
+  @Override
+  protected void setupModel() {
+    super.setupModel();
+
+    config = readConfig(model);
+    // TODO (garrettjones) depend on the framework to take care of this.
+    if (model.getErrorCount() > 0) {
+      for (Diag diag : model.getDiags()) {
+        System.err.println(diag.toString());
+      }
+      throw new IllegalArgumentException("Problem creating Generator");
+    }
+  }
+
+  @Override
+  protected boolean suppressDiagnosis() {
+    // Suppress linter warnings
+    return true;
+  }
+
+  @Override
+  protected String baselineFileName() {
+    String methodName = testName.getMethodName();
+    Matcher m = BASELINE_PATTERN.matcher(methodName);
+    if (m.find()) {
+      return m.group(2) + "_" + m.group(1) + ".baseline";
+    } else {
+      return name + "_" + methodName + ".baseline";
+    }
+  }
+
+  private ConfigProto readConfig(DiagCollector diagCollector) {
+    List<String> inputNames = new ArrayList<>();
+    List<String> inputs = new ArrayList<>();
+
+    for (String gapicConfigFileName : gapicConfigFileNames) {
+      URL gapicConfigUrl = getTestDataLocator().findTestData(gapicConfigFileName);
+      String configData = getTestDataLocator().readTestData(gapicConfigUrl);
+      inputNames.add(gapicConfigFileName);
+      inputs.add(configData);
+    }
+
+    ImmutableMap<String, Message> supportedConfigTypes =
+        ImmutableMap.<String, Message>of(
+            ConfigProto.getDescriptor().getFullName(), ConfigProto.getDefaultInstance());
+    ConfigProto configProto =
+        (ConfigProto) MultiYamlReader.read(model, inputNames, inputs, supportedConfigTypes);
+
+    return configProto;
+  }
+
+  public static List<Object[]> createTestedConfigs(String idForFactory, String[] gapicConfigFileNames) {
```

Comment:
Could you add a brief description of this method? The auto-creation of multiple test setups is a little magical, so it would be helpful to note that this is doing the work, and to give some sense of the structure of the return type.